src/OpenSpaceToolkit/Astrodynamics/Dynamics/Tabulated.cpp

```diff
@@ -0,0 +1,149 @@
+
+Tabulated::Tabulated(
+    const Array<Instant>& anInstantArray,
+    const MatrixXd& aContributionProfile,
+    const Array<Shared<const CoordinatesSubset>>& aWriteCoordinatesSubsets
+)
+    : Dynamics("Tabulated"),
+      contributionProfile_(aContributionProfile),
+      instants_(anInstantArray),
+      writeCoordinatesSubsets_(aWriteCoordinatesSubsets)
+{
+    if (anInstantArray.getSize() != (Index)aContributionProfile.rows())
+    {
+        throw ostk::core::error::RuntimeError(
+            "Contribution profile must have the same number of rows as the number of instants."
+        );
+    }
+
+    if (aWriteCoordinatesSubsets
+            .map<Index>(
+                [](const auto& coordinatesSubset)
+                {
+                    return coordinatesSubset->getSize();
+                }
+            )
+            .reduce(std::plus<Index>()) != (Index)aContributionProfile.cols())
+    {
+        throw ostk::core::error::RuntimeError(
+            "Contribution profile must have the same number of columns as the sum of the sizes of the write "
+            "coordinates subsets."
+        );
+    }
+
+    VectorXd timestamps(anInstantArray.getSize());
+    Index i = 0;
+    for (const auto& instant : anInstantArray)
+    {
+        timestamps(i) = (instant - Instant::J2000()).inSeconds();
+        ++i;
+    }
+
+    interpolators_.reserve(aContributionProfile.cols());
+    for (i = 0; i < (Index)aContributionProfile.cols(); ++i)
+    {
+        // TBI: In the future we can allow the user to specify the interpolator type.
+        interpolators_.add(BarycentricRational(timestamps, aContributionProfile.col(i)));
+    }
+}
+
+std::ostream& operator<<(std::ostream& anOutputStream, const Tabulated& aTabulated)
+{
+    aTabulated.print(anOutputStream);
+
+    return anOutputStream;
+}
+
+const Array<Instant>& Tabulated::accessInstants() const
+{
+    return instants_;
+}
+
+Array<Instant> Tabulated::getInstants() const
+{
+    return accessInstants();
+}
+
+const MatrixXd& Tabulated::accessContributionProfile() const
+{
+    return contributionProfile_;
+}
+
+MatrixXd Tabulated::getContributionProfile() const
+{
+    return accessContributionProfile();
+}
+
+bool Tabulated::isDefined() const
+{
+    return true;
+}
+
+Array<Shared<const CoordinatesSubset>> Tabulated::getReadCoordinatesSubsets() const
+{
+    return {};
+}
+
+Array<Shared<const CoordinatesSubset>> Tabulated::getWriteCoordinatesSubsets() const
+{
+    return writeCoordinatesSubsets_;
+}
+
+VectorXd Tabulated::computeContribution(
+    const Instant& anInstant, [[maybe_unused]] const VectorXd& x, [[maybe_unused]] const Shared<const Frame>& aFrameSPtr
+) const
+{
+    const double epoch = (anInstant - Instant::J2000()).inSeconds();
+
+    VectorXd contribution(interpolators_.getSize());
+    for (Index i = 0; i < interpolators_.getSize(); ++i)
+    {
+        contribution(i) = interpolators_[i].evaluate(epoch);
+    }
+
+    return contribution;
```

Comment:
Frame should not be ignored here.

I am not suggesting we go crazy either. Some ideas...

### "Let's at least check and crash"

This option simply fails if Frames are not consistent.

```cpp
if (frameSPtr_ != aFrameSPtr) {
    // throw an error
}

return contribution;
```

### The "let's go crazy" option
This would involve understanding what the contribution is. Which we could do, as we got the `writeCoordinatesSubests` for it...

Not so trivial as `writeCoordinatesSubsets` can be anything. 
- If it's a `CartesianVelocity` (i.e. an acceleration contribution) then one could just rotate the vector using `aFrameSPtr.getTransformTo(frameSPtr_)`

Not even sure about that second option, but bottomline is we should a) have a frame attached to the tabulated profile b) at least fail if it's not the same as the solver is expected when calling `computeContribution`

----

### Note

I know I mentioned earlier that `Tabulated` should not talk about `accelerations` as you made it a generic class, which is great. I am just thinking IF having a structure like this would make the whole frame business easier:

```
Dynamics
  Tabulated.hpp 
      - Handles timestampes, matrixes, interpolation, etc.
      - `computeContribution` calls `transformContribution(contribution)` at the end
      - Default `transformContribution` simply checks if frames are consistent, fails otherwise
```

Then you can have children classes, like a `TabulatedManeuver` which contributes to `CartesianVelocity` and `MassDerivatives`:

```
Dynamics
  Tabulated
    TabulatedManeuver.hpp
        - Extends `Tabulated.hpp`
        - Assumes cols=4 being `CartesianVelocity` and `Mass`
        - Overrides `transformContribution` function to simply rotate the `CartesianVelocity` contribution and attach the unchanged mass contribution
```

Again... just food for thought 😊

Reply:
This is an excellent point. I've implement idea 1. I think after a team discussion on how to handle subsetDerivatives we can implement idea 2 but having some smart static constructors.